tools/evict-shared-cache/evict-shared-cache.cpp

```diff
@@ -49,11 +50,50 @@ std::unique_ptr<EvictionPolicy> make_policy(const char* argv0, const char* polic
     return std::make_unique<NilEvictionPolicy>();
   }
 
-  std::cout << "Unknown policy: " << policy << std::endl;
+  std::cerr << "Unknown policy: " << policy << std::endl;
   print_help(argv0);
   exit(EXIT_FAILURE);
 }
 
+std::string read_command() {
+  std::string command = "";
+
+  // TODO: no idea if this is the right condition
+  while (std::cin.good() && std::cout.good()) {
+    uint8_t buffer[4096] = {};
+    ssize_t count = read(0, static_cast<void*>(buffer), 4096);
+
+    // Nothing new to process
+    if (count == 0) {
+      continue;
+    }
+
+    // An error occured during read
+    if (count < 0) {
+      std::cout << "Failed to read from stdin: " << strerror(errno) << std::endl;
+      exit(EXIT_FAILURE);
+    }
+
+    for (ssize_t i = 0; i < count; i++) {
+      char c = buffer[i];
+      if (c != '\0') {
+        command += c;
+        continue;
+      }
+
+      if (c != i - 1) {
+        std::cerr << "Multiple commands in one read. Trailing command will be dropped."
+                  << std::endl;
+      }
+
+      return command;
+    }
```

Comment:
```
std::vector<std::string> commands;
....
uint8_t* iter = buffer;
const uint8_t* buffer_end = buffer + count;
while (iter < buffer_end) {
    auto end = std::find(iter, buffer_end, '\0');
    command.append(iter, end);
    if (end != buffer_end) {
      commands.emplace_back(std::move(command));
    }
    iter = end;
}
```

to avoid loops and to handle batched commands we can do something like this instead.

Reply:
What is the point of only `emplace_back`ing inside of  `if (end != buffer_end)` here?

`end` always signifies the end of a command so shouldn't it always emplace iter...end as long as '\0' was found?

Reply:
end != buffer_end is exactly the condition you need to check for to see if you found a null terminator or not. If end == buffer_end that means there was no null terminator so the next read will have to fill the buffer.

As another thought. This is going to have to be a class method so that command can persist across invocations.